MessageCorrelationTest is unstable #2

@saig0, found another one while testing the fix for #1234:

```
[ERROR] Tests run: 30, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 64.807 s <<< FAILURE! - in io.zeebe.broker.workflow.MessageCorrelationTest
[ERROR] shouldCorrelateMessageByCorrelationKey[intermediate message catch event](io.zeebe.broker.workflow.MessageCorrelationTest)  Time elapsed: 11.626 s  <<< ERROR!
java.lang.RuntimeException: no more events available
        at io.zeebe.test.broker.protocol.clientapi.RawMessageCollector.get(RawMessageCollector.java:51)
        at io.zeebe.test.broker.protocol.clientapi.RawMessageCollector.get(RawMessageCollector.java:25)
        at java.util.stream.StreamSpliterators$InfiniteSupplyingSpliterator$OfRef.tryAdvance(StreamSpliterators.java:1356)
        at java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:126)
        at java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:498)
        at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:485)
        at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
        at java.util.stream.FindOps$FindOp.evaluateSequential(FindOps.java:152)
        at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.util.stream.ReferencePipeline.findFirst(ReferencePipeline.java:464)
        at io.zeebe.test.util.stream.StreamWrapper.findFirst(StreamWrapper.java:261)
        at io.zeebe.test.util.stream.StreamWrapper.findFirst(StreamWrapper.java:261)
        at io.zeebe.test.util.stream.StreamWrapper.findFirst(StreamWrapper.java:261)
        at io.zeebe.test.util.stream.StreamWrapper.findFirst(StreamWrapper.java:261)
        at io.zeebe.test.util.stream.StreamWrapper.findFirst(StreamWrapper.java:261)
        at io.zeebe.test.broker.protocol.clientapi.TestPartitionClient.receiveFirstWorkflowInstanceEvent(TestPartitionClient.java:483)
        at io.zeebe.broker.workflow.MessageCorrelationTest.shouldCorrelateMessageByCorrelationKey(MessageCorrelationTest.java:365)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
        at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
        at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
        at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
        at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
        at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
        at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:55)
        at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:48)
        at org.junit.rules.RunRules.evaluate(RunRules.java:20)
        at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
        at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
        at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
        at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
        at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
        at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
        at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
        at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
        at org.junit.runners.Suite.runChild(Suite.java:128)
        at org.junit.runners.Suite.runChild(Suite.java:27)
        at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
        at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
        at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
        at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
        at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
        at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
        at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:365)
        at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:273)
        at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:238)
        at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:159)
        at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:383)
        at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:344)
        at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:125)
        at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:417)
```

--- a/broker-core/src/main/java/io/zeebe/broker/subscription/command/SubscriptionApiCommandMessageHandler.java
+++ b/broker-core/src/main/java/io/zeebe/broker/subscription/command/SubscriptionApiCommandMessageHandler.java
@@ -165,6 +165,7 @@ public class SubscriptionApiCommandMessageHandler implements ServerMessageHandle
   private boolean onCorrelateMessageSubscription(DirectBuffer buffer, int offset, int length) {
     correlateMessageSubscriptionCommand.wrap(buffer, offset, length);
 
+    messageSubscriptionRecord.reset();
     messageSubscriptionRecord
         .setWorkflowInstancePartitionId(
             openMessageSubscriptionCommand.getWorkflowInstancePartitionId())

--- a/broker-core/src/main/java/io/zeebe/broker/subscription/message/data/MessageSubscriptionRecord.java
+++ b/broker-core/src/main/java/io/zeebe/broker/subscription/message/data/MessageSubscriptionRecord.java
@@ -30,7 +30,7 @@ public class MessageSubscriptionRecord extends UnpackedObject {
   private final LongProperty workflowInstanceKeyProp = new LongProperty("workflowInstanceKey");
   private final LongProperty activityInstanceKeyProp = new LongProperty("activityInstanceKey");
   private final StringProperty messageNameProp = new StringProperty("messageName");
-  private final StringProperty correlationKeyProp = new StringProperty("correlationKey");
+  private final StringProperty correlationKeyProp = new StringProperty("correlationKey", "");
 
   public MessageSubscriptionRecord() {
     this.declareProperty(workflowInstancePartitionIdProp)

--- a/broker-core/src/main/java/io/zeebe/broker/subscription/message/state/SubscriptionState.java
+++ b/broker-core/src/main/java/io/zeebe/broker/subscription/message/state/SubscriptionState.java
@@ -104,7 +104,7 @@ public class SubscriptionState<T extends Subscription> {
     return getSubscription(keyBuffer, KEY_OFFSET, subscription.getKeyLength());
   }
 
-  public T getSubscription(final DirectBuffer buffer, final int offset, final int length) {
+  private T getSubscription(final DirectBuffer buffer, final int offset, final int length) {
     return persistenceHelper.getValueInstance(
         clazz, subscriptionHandle, buffer, offset, length, valueBuffer);
   }
@@ -112,7 +112,8 @@ public class SubscriptionState<T extends Subscription> {
   public List<T> findSubscriptions(
       final DirectBuffer messageName, final DirectBuffer correlationKey) {
     final List<T> subscriptionsList = new ArrayList<>();
-    rocksDbWrapper.whileTrue(
+
+    rocksDbWrapper.foreach(
         subscriptionHandle,
         (key, value) -> {
           try {
@@ -125,11 +126,11 @@ public class SubscriptionState<T extends Subscription> {
             if (isEqual) {
               subscriptionsList.add(subscription);
             }
-            return isEqual;
           } catch (Exception ex) {
             throw new RuntimeException(ex);
           }
         });
+
     return subscriptionsList;
   }
 

--- a/broker-core/src/test/java/io/zeebe/broker/subscription/message/state/MessageStateControllerTest.java
+++ b/broker-core/src/test/java/io/zeebe/broker/subscription/message/state/MessageStateControllerTest.java
@@ -178,8 +178,7 @@ public class MessageStateControllerTest {
         new MessageSubscription(
             "messageName", "correlationKey", "{\"foo\":\"bar\"}", 2, 2, 3, 1234);
     final MessageSubscription subscription3 =
-        new MessageSubscription(
-            "messageName", "correlationKey", "{\"foo\":\"bar\"}", 3, 2, 3, 1234);
+        new MessageSubscription("otherName", "correlationKey", "{\"foo\":\"bar\"}", 3, 2, 3, 1234);
     stateController.put(subscription);
     stateController.put(subscription2);
     stateController.put(subscription3);
@@ -189,14 +188,19 @@ public class MessageStateControllerTest {
         stateController.findSubscriptions(wrapString("messageName"), wrapString("correlationKey"));
 
     // then
-    assertThat(readSubscriptions.size()).isEqualTo(3);
+    assertThat(readSubscriptions.size()).isEqualTo(2);
 
     MessageSubscription readSubscription = readSubscriptions.get(0);
     assertSubscription(subscription, readSubscription, 1234, 1, 2, 3);
     readSubscription = readSubscriptions.get(1);
-    assertSubscription(subscription, readSubscription, 1234, 2, 2, 3);
-    readSubscription = readSubscriptions.get(2);
-    assertSubscription(subscription, readSubscription, 1234, 3, 2, 3);
+    assertSubscription(subscription2, readSubscription, 1234, 2, 2, 3);
+
+    // and
+    final List<MessageSubscription> otherSubscriptions =
+        stateController.findSubscriptions(wrapString("otherName"), wrapString("correlationKey"));
+
+    assertThat(otherSubscriptions.size()).isEqualTo(1);
+    assertSubscription(subscription3, otherSubscriptions.get(0), 1234, 3, 2, 3);
   }
 
   @Test

--- a/broker-core/src/test/java/io/zeebe/broker/workflow/MessageCorrelationTest.java
+++ b/broker-core/src/test/java/io/zeebe/broker/workflow/MessageCorrelationTest.java
@@ -458,7 +458,7 @@ public class MessageCorrelationTest {
             entry("workflowInstanceKey", workflowInstanceKey),
             entry("activityInstanceKey", catchEventEntered.key()),
             entry("messageName", "order canceled"),
-            entry("correlationKey", "order-123"));
+            entry("correlationKey", ""));
   }
 
   @Test

